streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBRangeIteratorTest.java

```diff
@@ -0,0 +1,295 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.junit.Test;
+
+import static org.easymock.EasyMock.expect;
+import static org.easymock.EasyMock.partialMockBuilder;
+import static org.easymock.EasyMock.replay;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.core.Is.is;
+
+public class RocksDBRangeIteratorTest {
+
+    private RocksDBRangeIterator mockRocksDBRangeIterator() {
+        return partialMockBuilder(RocksDBRangeIterator.class)
+                .addMockedMethod("getNext")
+                .addMockedMethod("getRawLastKey")
+                .addMockedMethod("isForward")
+                .addMockedMethod("getComparator")
+                .createMock();
+    }
+
+    @Test
+    public void shouldReturnAllKeysInTheRangeInForwardDirection() {
+        final RocksDBRangeIterator rocksDBRangeIterator = mockRocksDBRangeIterator();
+
+        final String key1 = "a";
+        final String key2 = "b";
+        final String key3 = "c";
+        final String key4 = "d";
+        final String value = "value";
+
+        final  Bytes key1Bytes = Bytes.wrap(key1.getBytes());
+        final  Bytes key2Bytes = Bytes.wrap(key2.getBytes());
+        final  Bytes key3Bytes = Bytes.wrap(key3.getBytes());
+        final  Bytes key4Bytes = Bytes.wrap(key4.getBytes());
+        final byte[] valueBytes = value.getBytes();
+
+        expect(rocksDBRangeIterator.getNext())
+            .andReturn(KeyValue.pair(key1Bytes, valueBytes))
+            .andReturn(KeyValue.pair(key2Bytes, valueBytes))
+            .andReturn(KeyValue.pair(key3Bytes, valueBytes))
+            .andReturn(KeyValue.pair(key4Bytes, valueBytes));
+
+        expect(rocksDBRangeIterator.getRawLastKey())
+            .andReturn(key3Bytes.get())
+            .anyTimes();
+
+        expect(rocksDBRangeIterator.isForward())
+            .andReturn(true)
+            .anyTimes();
+
+        expect(rocksDBRangeIterator.getComparator())
+            .andReturn(Bytes.BYTES_LEXICO_COMPARATOR)
+            .anyTimes();
+
+
+        replay(rocksDBRangeIterator);
+        assertThat(rocksDBRangeIterator.makeNext().key, is(key1Bytes));
+        assertThat(rocksDBRangeIterator.makeNext().key, is(key2Bytes));
+        assertThat(rocksDBRangeIterator.makeNext().key, is(key3Bytes));
+        assertThat(rocksDBRangeIterator.makeNext(), is(nullValue()));
```

Comment:
You usually do not want to mock the class under test, because you want to test it. Also partial mocks should only be used if absolutely necessary. A rule of thumb is if a partial mock is needed then most probably the design has a flaw.
In this specific case, you should mock RocksDB's iterator. For the class under test, you should test `hasNext()`, `next()`, `peekNextKey()` and `close()`, because those are the one exposed (`makeNext()` should actually de declared as `protected`, IMO). 

```suggestion
        final String key1 = "a";
        final String key2 = "b";
        final String key3 = "c";
        final String key4 = "d";
        final String value = "value";

        final  Bytes key1Bytes = Bytes.wrap(key1.getBytes());
        final  Bytes key2Bytes = Bytes.wrap(key2.getBytes());
        final  Bytes key3Bytes = Bytes.wrap(key3.getBytes());
        final  Bytes key4Bytes = Bytes.wrap(key4.getBytes());
        final byte[] valueBytes = value.getBytes();

        final RocksIterator rocksIterator = mock(RocksIterator.class);
        rocksIterator.seek(key1Bytes.get());
        expect(rocksIterator.isValid())
            .andReturn(true)
            .andReturn(true)
            .andReturn(true)
            .andReturn(true)
            .andReturn(false);
        expect(rocksIterator.key())
            .andReturn(key1Bytes.get())
            .andReturn(key2Bytes.get())
            .andReturn(key3Bytes.get())
            .andReturn(key4Bytes.get());
        expect(rocksIterator.value()).andReturn(valueBytes).times(4);
        rocksIterator.next();
        expectLastCall().times(4);
        replay(rocksIterator);

        final RocksDBRangeIterator rocksDBRangeIterator = new RocksDBRangeIterator(
            STORE_NAME,
            rocksIterator,
            Collections.emptySet(),
            key1Bytes,
            key4Bytes,
            true
        );
        assertThat(rocksDBRangeIterator.hasNext(), is(true));
        assertThat(rocksDBRangeIterator.next().key, is(key1Bytes));
        assertThat(rocksDBRangeIterator.hasNext(), is(true));
        assertThat(rocksDBRangeIterator.next().key, is(key2Bytes));
        assertThat(rocksDBRangeIterator.hasNext(), is(true));
        assertThat(rocksDBRangeIterator.next().key, is(key3Bytes));
        assertThat(rocksDBRangeIterator.hasNext(), is(true));
        assertThat(rocksDBRangeIterator.next().key, is(key4Bytes));
        assertThat(rocksDBRangeIterator.hasNext(), is(false));
        verify(rocksIterator);
```

Reply:
@cadonna , honestly even I wasn't making the changes that I ended up making but I assumed that makeNext is the only method that needs to be tested wrt Range Iterator as other methods are inherited by it so those should have test cases separately. I have changes the test cases now.

Reply:
@cadonna , did you get  a chance to review these changes?